### ISSUE TITLE
模型导出与不导出时预测的class_id差1

### ISSUE THREAD

**Issue by xzkzdx** [PaddlePaddle/PaddleDetection#3764]
不导出模型的时候使用tools/infer.py识别的物体class_id正确，
python tools/export_model.py导出模型后，使用python deploy/python/infer.py --model_dir=模型导出位置发现class_id少1

**Comment by yghstill** [comment]
@xzkzdx 哪个模型？配置文件改过吗？

**Comment by xzkzdx** [comment]
cascade_rcnn_cls_aware_r200_vd_fpn_dcnv2_nonlocal_softnms.yml
architecture: CascadeRCNNClsAware
max_iters: 100000
snapshot_iter: 100000
use_gpu: true
log_iter: 100
save_dir: output
pretrain_weights: https://paddle-imagenet-models-name.bj.bcebos.com/ResNet200_vd_pretrained.tar
weights: output/cascade_rcnn_cls_aware_r200_vd_fpn_dcnv2_nonlocal_softnms/model_final
# obj365 dataset format and its eval method are same as those for coco
metric: COCO
num_classes: 366

CascadeRCNNClsAware:
  backbone: ResNet
  fpn: FPN
  rpn_head: FPNRPNHead
  roi_extractor: FPNRoIAlign
  bbox_head: CascadeBBoxHead
  bbox_assigner: CascadeBBoxAssigner

ResNet:
  norm_type: bn
  depth: 200
  feature_maps: [2, 3, 4, 5]
  freeze_at: 2
  variant: d
  dcn_v2_stages: [3, 4, 5]
  nonlocal_stages: [4]

FPN:
  min_level: 2
  max_level: 6
  num_chan: 256
  spatial_scale: [0.03125, 0.0625, 0.125, 0.25]

FPNRPNHead:
  anchor_generator:
    anchor_sizes: [32, 64, 128, 256, 512]
    aspect_ratios: [0.5, 1.0, 2.0]
    stride: [16.0, 16.0]
    variance: [1.0, 1.0, 1.0, 1.0]
  anchor_start_size: 32
  min_level: 2
  max_level: 6
  num_chan: 256
  rpn_target_assign:
    rpn_batch_size_per_im: 256
    rpn_fg_fraction: 0.5
    rpn_positive_overlap: 0.7
    rpn_negative_overlap: 0.3
    rpn_straddle_thresh: 0.0
  train_proposal:
    min_size: 0.0
    nms_thresh: 0.7
    pre_nms_top_n: 2000
    post_nms_top_n: 2000
  test_proposal:
    min_size: 0.0
    nms_thresh: 0.7
    pre_nms_top_n: 1000
    post_nms_top_n: 1000

FPNRoIAlign:
  canconical_level: 4
  canonical_size: 224
  min_level: 2
  max_level: 5
  box_resolution: 14
  sampling_ratio: 2

CascadeBBoxAssigner:
  batch_size_per_im: 512
  bbox_reg_weights: [10, 20, 30]
  bg_thresh_lo: [0.0, 0.0, 0.0]
  bg_thresh_hi: [0.5, 0.6, 0.7]
  fg_thresh: [0.5, 0.6, 0.7]
  fg_fraction: 0.25
  class_aware: True

CascadeBBoxHead:
  head: CascadeTwoFCHead
  #nms: MultiClassSoftNMS
  nms: MultiClassNMS

CascadeTwoFCHead:
  mlp_dim: 1024

MultiClassSoftNMS:
  score_threshold: 0.001
  keep_top_k: 300
  softnms_sigma: 0.15

LearningRate:
  base_lr: 0.0001
  schedulers:
  - !PiecewiseDecay
    gamma: 0.1
    milestones: [520000, 740000]
  - !LinearWarmup
    start_factor: 0.1
    steps: 1000

OptimizerBuilder:
  optimizer:
    momentum: 0.9
    type: Momentum
  regularizer:
    factor: 0.0001
    type: L2

TrainReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'gt_bbox', 'gt_class', 'is_crowd']
  dataset:
    !COCODataSet
    dataset_dir: /export/dataset/obj365
    anno_path: train.json
    image_dir: train
  sample_transforms:
  - !DecodeImage
    to_rgb: True
  - !RandomFlipImage
    prob: 0.5
  - !NormalizeImage
    is_channel_first: false
    is_scale: True
    mean:
    - 0.485
    - 0.456
    - 0.406
    std:
    - 0.229
    - 0.224
    - 0.225
  - !ResizeImage
    interp: 1
    target_size: [416, 448, 480, 512, 544, 576, 608, 640, 672, 704, 736, 768, 800, 832, 864, 896, 928, 960, 992, 1024, 1056, 1088, 1120, 1152, 1184, 1216, 1248, 1280, 1312, 1344, 1376, 1408]
    max_size: 1800
    use_cv2: true
  - !Permute
    to_bgr: false
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
  batch_size: 1
  shuffle: true
  drop_last: false
  worker_num: 1
  use_process: false

EvalReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'im_shape']
  dataset:
    !COCODataSet
    dataset_dir: /export/dataset/obj365
    anno_path: val.json
    image_dir: val
  sample_transforms:
  - !DecodeImage
    to_rgb: True
    with_mixup: False
  - !NormalizeImage
    is_channel_first: false
    is_scale: True
    mean:
    - 0.485
    - 0.456
    - 0.406
    std:
    - 0.229
    - 0.224
    - 0.225
  - !ResizeImage
    interp: 1
    target_size:
    - 1200
    max_size: 2000
    use_cv2: true
  - !Permute
    to_bgr: false
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
  batch_size: 1
  worker_num: 2
  drop_empty: false

TestReader:
  inputs_def:
    fields: ['image', 'im_info', 'im_id', 'im_shape']
  dataset:
    !ImageFolder
    dataset_dir: dataset/obj365/annotations
    anno_path: annotations.json
    #anno_path: annotations/instances_val2017.json
  sample_transforms:
  - !DecodeImage
    to_rgb: true
    with_mixup: false
  - !NormalizeImage
    is_channel_first: false
    is_scale: true
    mean: [0.485,0.456,0.406]
    std: [0.229, 0.224,0.225]
  - !ResizeImage
    interp: 1
    max_size: 1333
    target_size: 800
    use_cv2: true
  - !Permute
    channel_first: true
    to_bgr: false
  batch_transforms:
  - !PadBatch
    pad_to_stride: 32
    use_padded_im_info: true
  batch_size: 1
  worker_num: 1

## 导出的yml文件：
#### inference_model/cascade_rcnn_cls_aware_r200_vd_fpn_dcnv2_nonlocal_softnms/infer_cfg.yml
use_python_inference: false
mode: fluid
draw_threshold: 0.5
metric: COCO
arch: RCNN
min_subgraph_size: 40
with_background: true
Preprocess:
- is_channel_first: false
  is_scale: true
  mean:
  - 0.485
  - 0.456
  - 0.406
  std:
  - 0.229
  - 0.224
  - 0.225
  type: Normalize
- interp: 1
  max_size: 1333
  resize_box: false
  target_size: 800
  type: Resize
  use_cv2: true
- channel_first: true
  to_bgr: false
  type: Permute
- stride: 32
  type: PadStride
label_list:
- ship
- okra
- printer
- cigar
- cellphone
- mango
- pillow
- billiards
- iron
- bench
- fork
- hat
- bear
- avocado
- red cabbage
- stapler
- cup
- flute
- hamburger
- plum
- luggage
- knife
- airplane
- toiletries
- starfish
- pine apple
- wild bird
- zebra
- skating and skiing shoes
- yak
- sandwich
- Life saver
- scissors
- clock
- desk
- asparagus
- duck
- cabinet/shelf
- carriage
- board eraser
- golf ball
- polar bear
- eggplant
- tape measure/ruler
- leather shoes
- speaker
- butterfly
- skis
- microphone
- belt
- tablet
- spring rolls
- mirror
- oven
- key
- radiator
- slide
- swan
- monitor
- goose
- shovel
- sushi
- barrel/bucket
- broom
- hamimelon
- egg
- CD
- mushroom
- van
- coffee table
- folder
- computer box
- watermelon
- storage box
- flashlight
- clutch
- urinal
- game board
- rice
- pomelo
- garlic
- meat balls
- wine glass
- flower
- picture/frame
- corn
- sports car
- hair dryer
- hot dog
- necklace
- trash bin/can
- coffee machine
- hurdle
- pot/pan
- laptop
- saw
- basketball
- american football
- umbrella
- penguin
- kettle
- nuts
- tennis racket
- paper towel
- ring
- book
- motorbike/motorcycle
- apple
- suv
- dumbbell
- green onion
- binoculars
- sausage
- swing
- eraser
- rice cooker
- eagle
- bowl
- tie
- fire extinguisher
- bus
- lobster
- pickup truck
- fire truck
- lantern
- onion
- toilet
- bow tie
- facial mask
- sneakers
- bathtub
- projector
- combine with bowl
- baseball bat
- trombone
- tripod
- broccoli
- pencil case
- sailboat
- dining table
- hammer
- chopsticks
- canned
- goldfish
- rabbit
- peach
- pasta
- remote
- lamp
- toothbrush
- paint brush
- basket
- traffic light
- refrigerator
- jellyfish
- carpet
- pliers
- Electronic stove and gas stove
- candle
- paddle
- microscope
- baseball
- blackboard/whiteboard
- kiwi fruit
- fishing rod
- backpack
- screwdriver
- 'tissue '
- cat
- cherry
- antelope
- extension cord
- lettuce
- seal
- gun
- green vegetables
- papaya
- hockey stick
- camel
- elephant
- toothpaste
- tomato
- steak
- train
- microwave
- cello
- plants pot/vase
- lighter
- fire hydrant
- drum
- calculator
- tricycle
- hoverboard
- potato
- glove
- bread
- liquid soap
- snowboard
- tangerine
- ladder
- surveillance camera
- handbag
- strawberry
- crane
- air conditioner
- poker card
- tortoise/turtle
- cow
- blender
- stuffed animal
- pie
- donkey
- tong
- towel/napkin
- toy
- shampoo/shower gel
- egg tart
- frisbee
- chicken
- shark
- carrot
- tent
- wheelchair
- sandals
- converter
- bed
- rickshaw
- lion
- cymbal
- combine with glove
- golf club
- ice cream
- kite
- french horn
- cheese
- lemon
- Bread/bun
- pigeon
- trolley
- fish
- volleyball
- balloon
- tennis ball
- target
- pen/pencil
- router/modem
- dog
- llama
- coconut
- monkey
- glasses
- keyboard
- cleaning products
- hotair balloon
- truck
- french fries
- human
- telephone
- cue
- pepper
- street lights
- durian
- Accordion/keyboard/piano
- noodles
- pomegranate
- stroller
- bottle
- fan
- chips
- chair
- mop
- scallop
- tape
- power outlet
- dolphin
- globe
- pig
- gas stove
- crosswalk sign
- surfboard
- marker
- vent
- camera
- boat
- saxophone
- toaster
- candy
- cake
- suitcase
- skateboard
- trophy
- pitaya
- potted plant
- pumpkin
- tuba
- washing machine
- parrot
- car
- radish
- race car
- pear
- horse
- sofa
- watch
- deer
- soccer
- plate
- orange
- mouse
- flag
- helicopter
- measuring cup
- table tennis  paddle
- radio
- giraffe
- grapes
- scale
- violin
- notepaper
- chainsaw
- sink
- cabbage
- faucet
- shrimp
- bar soap
- bracelet
- tea pot
- green beans
- boots
- stool
- guitar
- mangosteen
- donut
- dates
- slippers
- helmet
- electric drill
- pizza
- comb
- spoon
- bicycle
- medal
- traffic sign
- traffic cone
- horn
- hanger
- baozi
- machinery vehicle
- sheep
- treadmill
- cucumber
- nightstand
- earphone
- head phone
- ambulance
- banana
- brush
- barbell
- facial cleanser
- crab
- parking meter
- cutting/chopping board
- cookies
- trumpet
- high heels
- background

**Comment by yghstill** [comment]
@xzkzdx 将导出的infer_cfg.yml中label_list中的最后background放在第一位，再试下，这个问题我们跟进排查下。

**Comment by xzkzdx** [comment]
怎样配置多卡预测，文档里找不到

**Comment by houxiaonan** [comment]
> @xzkzdx 将导出的infer_cfg.yml中label_list中的最后background放在第一位，再试下，这个问题我们跟进排查下。

我也遇到同样的问题，麻烦帮忙解答一下，多谢。
